ikarus/finiteelements/mechanics/materials/vanishingstrain.hh

```diff
@@ -0,0 +1,180 @@
+ * \ingroup materials
+ * \tparam strainIndexPair An array of MatrixIndexPair representing fixed stress components.
+ * \tparam MI The underlying material model.
+ */
+template <auto strainIndexPair, typename MI>
+struct VanishingStrain : public Material<VanishingStrain<strainIndexPair, MI>>
+{
+  using Underlying         = MI;                              ///< The underlying material type.
+  using ScalarType         = typename Underlying::ScalarType; ///< Scalar type.
+  using MaterialParameters = typename Underlying::MaterialParameters;
+
+  static constexpr auto fixedPairs        = strainIndexPair;                     ///< Array of fixed stress components.
+  static constexpr auto freeVoigtIndices  = createfreeVoigtIndices(fixedPairs);  ///< Free Voigt indices.
+  static constexpr auto fixedVoigtIndices = createFixedVoigtIndices(fixedPairs); ///< Fixed Voigt indices.
+  static constexpr auto freeStrains       = freeVoigtIndices.size();             ///< Number of free strains.
+
+  static constexpr auto strainTag          = Underlying::strainTag;          ///< Strain tag.
+  static constexpr auto stressTag          = Underlying::stressTag;          ///< Stress tag.
+  static constexpr auto tangentModuliTag   = Underlying::tangentModuliTag;   ///< Tangent moduli tag.
+  static constexpr bool energyAcceptsVoigt = Underlying::energyAcceptsVoigt; ///< Energy accepts Voigt notation.
+  static constexpr bool stressToVoigt      = true;                           ///< Stress to Voigt notation.
+  static constexpr bool stressAcceptsVoigt = true;                           ///< Stress accepts Voigt notation.
+  static constexpr bool moduliToVoigt      = true;                           ///< Moduli to Voigt notation.
+  static constexpr bool moduliAcceptsVoigt = true;                           ///< Moduli accepts Voigt notation.
+  static constexpr double derivativeFactor = Underlying::derivativeFactor;   ///< Derivative factor.
+
+  /**
+   * \brief Constructor for VanishingStrain.
+   * \param mat The underlying material model.
+   */
+  explicit VanishingStrain(MI mat)
+      : matImpl_{mat} {}
+
+  [[nodiscard]] constexpr static std::string nameImpl() noexcept {
+    auto matName = MI::name() + "_VanishingStrain(";
+    for (auto p : fixedPairs)
+      matName += "(" + std::to_string(p.row) + std::to_string(p.col) + ")";
+    matName += ")";
+    return matName;
+  }
+
+  /**
+   * \brief Returns the material parameters stored in the material
+   */
+  MaterialParameters materialParametersImpl() const { return matImpl_.materialParametersImpl(); }
+
+  /**
+   * \brief Computes the stored energy for the PlaneStrain material.
+   * \tparam Derived The derived type of the input matrix.
+   * \param Eraw The strain mesasure
+   * \return ScalarType The stored energy.
+   */
+  template <typename Derived>
+  ScalarType storedEnergyImpl(const Eigen::MatrixBase<Derived>& E) const {
```

Comment:
For the class VanishingStress, the input is always the full strains (in 3D 6 and in 2D 3) and only the output is modified and for the Voigt notation, the stresses entries are removed and the smaller sized vector is returned.

Here for non-linear material it should be possible to insert the reduced strains and not the full sized ones  (in 3D 6 and in 2D 3), because on element level then the user has to inflate his strain vector.

Additionally it is not fully clear which stresses should be returned.
Imagine, I have a simulation with plane strain condition. Therefore, everything happens in the x,y plane(0,1) indices.
For the element routines for energy,stress and moduli, a natural return would be [S_xx, S_yy,S_xy] because these are the only ones needed for the calculations, since the other stress provides no energy ,since it has no partner.
Anyway for post-processing one might be interested in the stresses anyway and wants to write out [S_zz] to vtk.

This is currently not possible, since the Voigt notation is also applied to the stresses.
One way would be to just request the non-Voigt matrix and return the S(2,2) component, which would be ok for me.
But we have to careful here that this interface is self-explainatory, since this might be confosing. Can you add some Latex comments like in the Arc-Length method or some code snippet which explain the size of objects which are strain arguments and stress and tangent moduli returned values for the special cases, here?